setup.go

```diff
@@ -0,0 +1,104 @@
+package ocp_dnsnameresolver
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/coredns/caddy"
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/plugin"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+)
+
+const pluginName = "ocp_dnsnameresolver"
+
+var log = clog.NewWithPlugin(pluginName)
+
+func init() { plugin.Register(pluginName, setup) }
+
+func setup(c *caddy.Controller) error {
+	resolver, err := resolverParse(c)
+	if err != nil {
+		return plugin.Error(pluginName, err)
+	}
+
+	onStart, onShut, err := resolver.initPlugin()
+	if err != nil {
+		return plugin.Error(pluginName, err)
+	}
+
+	if onStart != nil {
+		c.OnStartup(onStart)
+	}
+	if onShut != nil {
+		c.OnShutdown(onShut)
+	}
+
+	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {
+		resolver.Next = next
+		return resolver
+	})
+
+	return nil
+}
+
+func resolverParse(c *caddy.Controller) (*OCPDNSNameResolver, error) {
+	resolver := New()
+
+	i := 0
+	for c.Next() {
+		if i > 0 {
+			return nil, plugin.ErrOnce
+		}
+		i++
+
+		// There shouldn't be any more arguments.
+		if len(c.RemainingArgs()) != 0 {
+			return nil, c.ArgErr()
```

Comment:
The description of `ArgErr()` does say that another argument was expected but not found. However, the error message returned works for cases where the argument count is incorrect.

```go
// ArgErr returns an argument error, meaning that another
// argument was expected but not found. In other words,
// a line break or open curly brace was encountered instead of
// an argument.
func (d *Dispenser) ArgErr() error {
	if d.Val() == "{" {
		return d.Err("Unexpected token '{', expecting argument")
	}
	return d.Errf("Wrong argument count or unexpected line ending after '%s'", d.Val())
}
```

This error has also been used in other plugins. For example, in cache plugin it is used to indicate that the number of arguments should not be 0 or greater than 3: https://github.com/openshift/coredns/blob/022a0530038602605b8f3e8866c2a6ded97708cc/plugin/cache/setup.go#L137-L139

Reply:
Seems a little imprecise to me but it seems to be a generic argument error function which is commonly used. Closing this one and the similar ones.